### DISCUSSION TITLE
Syncing across multiple machines

### DISCUSSION THREAD

**Discussion by bencodezen** [johnlindquist/kit#577] [Q&A] [answered]
I often jump between multiple computers and it'd be great to keep the changes I make on one to another. Is there a recommended strategy for managing this via Dropbox / Github / some other solution?

**Comment by johnlindquist** [comment, marked as answer]
Great question!

Use standard git workflows on your `~/.kenv`. The only requirement is the path is `~/.kenv` on each machine.

## shared kenv

Using the "Manage kenv" menu, you can also create a "shared kenv" if you want to only share certain scripts.

## sharing a kenv

Once your kenv is up on Github, then you can send the url to a friend who can "Manage Kenv" -> clone your scripts so you can work on them together.

## p.s.

Would ❤️ to see a Netlify kenv 👀

**Reply by ApolloTang** [reply]
I think this is a bug:

ctr+; > kit > Manage Kenvs > New kenv 

will not create the folder with the provided name, instead, it created a folder called  "kenv-clone":

![image](https://user-images.githubusercontent.com/1150961/147373653-183bd7dd-8270-4b80-b14e-0b8f969028ef.png)

Despite that bug, to share my kit's scripts between computers via a GitHub repo, I think my next few step are:

1. rename the folder `~/.kenv/kenvs/kenv-clone/` to `~/.kenv/kenvs/my-kit-repo/`
2. add `~/.kenv/kenvs/my-kit-repos/bin/` to my bash or zsh's PATH variable
3. git init `~/.kenv/kenvs/my-kit-repo/`
4. normal git flow: git add, git commit etc. 

On another computer:
1. `cd ~/.kenv/kenvs/`
2. clone my repo
3. add `~/.kenv/kenvs/my-kit-repos/bin/` to my bash or zsh's PATH variable
4. normal git flow: git pull etc...

**Reply by ApolloTang** [reply]
oh.. looks like I am missing something.
executing the script in `~/.kenv/kenvs/my-kit-repos/bin/` from the terminal requires npm installation somewhere: 

![image](https://user-images.githubusercontent.com/1150961/147374092-885f0636-04ca-4fee-8ec9-7fd5165634df.png)

However, my script `foo` is callable from the kit app. 
my next question is how to call the script from the terminal?

**Reply by ApolloTang** [reply]
related: https://github.com/johnlindquist/kit/discussions/263

**Reply by johnlindquist** [reply]
Dang, sorry about the kenv create/clone bugs, I'll get that sorted for the next release.

## sharing kenvs

You're right, it's a standard git workflow:
1. Create a folder named `whatever` in `kenvs`
2. git init the `whatever` dir

From there, you can start using the app and it will generated the scripts/bin dirs.

## terminal/npm bug...

1. What is the contents of `foo.js`?
2. If you manually renamed the kenv after creating the script, you would also have to update the contents of the `bin/foo` file

This "regenerate bins" should help with that:
<img width="817" alt="CleanShot 2021-12-27 at 13 23 31@2x" src="https://user-images.githubusercontent.com/36073/147504878-34763655-67d6-4612-9a3e-ed9475947391.png">

Looks like you already did this, but this command will help run them from the terminal:

<img width="816" alt="CleanShot 2021-12-27 at 13 19 07@2x" src="https://user-images.githubusercontent.com/36073/147504589-70aac8d4-9a79-41a2-890e-cdaca17694f8.png">

**Reply by ApolloTang** [reply]
Hi thank you for answering my question.

the content of `foo` are as follow: 

```
import "@johnlindquist/kit"

say('hello')
```
The path of the above is `~/.kenv/kenvs/kenv-myscripts/scripts/foo.ts`

I also have an auto generated `foo` located in `bin` at this location:
```
file: ~/.kenv/kenvs/kenv-myscripts/bin/foo
```
with content: 
```
#!/bin/sh

NODE_NO_WARNINGS=1 \
KENV=$(cd "$(dirname ${BASH_SOURCE[0]})"/.. &> /dev/null && pwd) \
/Users/apollotang/.kit/node/bin/node \
/Users/apollotang/.kit/run/terminal.js \
/Users/apollotang/.kenv/kenvs/kenv-myscripts/scripts/foo.ts \
"$@"
```
Note that I had manually added `~/.kenv/kenvs/kenv-myscripts/bin` to my bash's PATH
That is how I can have `foo` available in my terminal: 

![image](https://user-images.githubusercontent.com/1150961/147526316-4b5bb03f-5837-46b0-b00e-e1889750d8bc.png)

However,  I am guessing... I shouldn't have running `foo` directly. 
I think I should have been executing the scripts in `bin` with the command `kit > run > foo`; but I got the following error: 

![image](https://user-images.githubusercontent.com/1150961/147526620-2db9ab36-5639-4fa6-ae42-14e0b3c0e6a5.png)

it say `foo not found`, but `foo` is one of the item in the list 

![image](https://user-images.githubusercontent.com/1150961/147526868-646d75ca-b1f5-4f89-95ae-f546a1e33559.png)

I admit I don't fully understand how the scripts in `bin` are accessed.

**Reply by ApolloTang** [reply]
Further insight 

If I create another script in `~/.kenv/scripts/foo3.ts` with this content :
```
// Name: foo3

import "@johnlindquist/kit"

say('foo three')
```
The error is ReferenceError related to `applescript`: 

![image](https://user-images.githubusercontent.com/1150961/147527449-19c1721e-ccfb-463f-b7b1-789d961a9e49.png)

So, it seem like there are two bugs: 
1. command `kit run` can not find script in `~/.kenv/kenvs/kenv-myscripts/bin` 
2. AppleScript is missing in `~/.kenv/bin`

**Reply by johnlindquist** [reply]
Looks like it's an issue when running from bin executables in TypeScript mode and how they're looking up their own kenv path. I'm running in JS mode, so I wasn't seeing the errors you're seeing. I'll get it fixed and put some tests around it.

A workaround, for now, would be replacing the contents of the bin file with:

```bash
#!/bin/sh

~/.kit/bin/kit \
~/.kenv/kenvs/kenv-myscripts/scripts/foo.ts \
"$@"
```

You can essentially pass any JS/TS file to `kit` and it will attempt to run it.

## bin explanation

In case it helps, here's what the bin file is doing:

1. `#!/bin/sh` - tells the terminal to run the code below with `sh`, the "shell command language". Bash and zsh are built on top of sh.
2. `~/.kit/bin/kit` - This file executes node with kit's setup scripts and environment variables then waits for your script to be passed in. If it doesn't receive your file, then it will start the "cli" version of kit.
3. `\` These backslashes are for "multiline" scripts. Otherwise the entire command would have to be on a single line
4. "$@" - This means to pass any additional arguments to the script. So if you do `kit my-script.js apple banana`, it will pass `apple` and `banana` as arguments to your script

**Comment by antenando** [comment]
We're in July 2022, ScriptKit is 1.20.3

I've been trying to sync ~/.kenv through multiple machines by cloning my .kenv repo and symlinking it to ~/.kenv, but I believe this is not the best approach, since for some reason ScriptKit removes `.git` folder from my .kenv. 

So, could you please enlighten us on the best way to share simple personal scripts we don't want to publicly publish?

Thanks for such an amazing app!

**Comment by johnlindquist** [comment]
@antenando Thanks for the kind words 😊

## Version 
The current version of Script Kit is 1.20.*5* (not .3). It should auto-update if you "check for updates" from the menubar.

## Sync a Main Kenv

1. quit Kit.app
2. rm existing `~/.kenv`
3. `git clone` your kenv to `~/.kenv`
4. open Kit.app

> A kenv's only requirement is that it contains a `scripts` directory. Kit.app will add the other directories as needed.

I'm going to dive back into the UX of kenvs. I know it's confusing and I'm super open to feedback!

## Share a Script without Publishing

I might be mis-reading your "share simple personal scripts we don't want to publicly publish", but here goes anyway: With a script selected, press right or cmd+k to open the share/options menu, select "link":

<img width="826" alt="CleanShot 2022-07-16 at 14 48 44@2x" src="https://user-images.githubusercontent.com/36073/179371529-64d7bebe-d3b0-44db-9abc-7302fc6bb66f.png">

This will create a "kit protocol" URL which you can share in Slack, etc (some apps like Discord, gists, and others don't like protocol URLs, so you'll have to use "copy" and copy/paste the entire script). The entire script is encorded into URL params like so:

`kit://snippet?name=aif-to-wav&content=Ly8gTmFtZTogQWlmIHRvIFdhdgoKaW1wb3J0ICJAam9obmxpbmRxdWlzdC9raXQiCgpsZXQgZmlsZXMgPSBhd2FpdCBnZXRTZWxlY3RlZEZpbGUoKQoKZm9yIGF3YWl0IChsZXQgYWlmIG9mIGZpbGVzLnNwbGl0KCJcbiIpKSB7CiAgICBsZXQgeyBkaXIsIG5hbWUgfSA9IHBhdGgucGFyc2UoYWlmKQogICAgbGV0IHdhdiA9IHBhdGguam9pbihkaXIsIG5hbWUgKyAiLndhdiIpCiAgICBhd2FpdCBleGVjKGBmZm1wZWcgLWkgJHthaWZ9ICR7d2F2fWApCn0KCgo`

### Bug? Deleting the .git dir?

1. Are you using the Kit.app "clone kenv" to pull down your repo?
2. Are you using the Kit.app "link kenv" to create the symlink?

I'll run some tests to see if I can figure out why it would be deleting the .git dir 🤔

Let me know if you have any other questions!

---
---
---

## Some other stuff I wrote that doesn't specifically address your questions but was on my mind

### Goal of kenvs

A big, future goal of Script Kit is to be able to run/manage any directory on your machine that contains a `scripts` directory. For example, adding a `~/Downloads/scripts` dir and writing scripts to manage your downloads. The same concept applies to image directories, web projects, etc. It's a complicated and challenging goal, but I think we can come up with something to make it all smooth process. Essentially, I don't want people to have to think about "kenvs" as much as "I want Script Kit to help me run scripts in this directory".

### Why ~/.kenv and ~/.kenv/kenvs

Kit.app has currently has the concept of a "main kenv":
~/.kenv/

and remote/other kenvs:
~/.kenv/kenvs

> For example, if you run `ls ~/.kenv/kenvs/` in the terminal, you should see an `examples` dir where the examples are stored.

This structure has mainly been for simplifying the file watcher for tracking changes to scripts in `~/.kenv/scripts/*` and `~/.kenv/kenvs/*/scripts/*`. I've recently made significant changes to the file watching for performance improvements, so I need to go back through the kenv add/remove/etc logic to ensure it didn't impact anything.
